http/php/guzzle/src/Middleware/CompressionHandler.php

```diff
@@ -0,0 +1,137 @@
+<?php
+/**
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.
+ * Licensed under the MIT License.  See License in the project root
+ * for license information.
+ */
+
+
+namespace Microsoft\Kiota\Http\Middleware;
+
+use GuzzleHttp\Promise\Create;
+use GuzzleHttp\Promise\PromiseInterface;
+use Microsoft\Kiota\Http\Middleware\Options\CompressionOption;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Class CompressionHandler
+ *
+ * Compresses a request body using the provided callbacks in {@link CompressionOption}
+ * Should the server return a 415, the CompressionHandler retries the request ONLY once with an uncompressed body
+ *
+ * @package Microsoft\Kiota\Http\Middleware
+ * @copyright 2022 Microsoft Corporation
+ * @license https://opensource.org/licenses/MIT MIT License
+ * @link https://developer.microsoft.com/graph
+ */
+class CompressionHandler
+{
+    private const COMPRESSION_RETRY_ATTEMPT = 'compressionRetryAttempt';
+
+    /**
+     * @var CompressionOption {@link CompressionOption}
+     */
+    private CompressionOption $compressionOption;
+
+    /**
+     * @var callable(RequestInterface, array): PromiseInterface
+     * Next handler to be called in the middleware pipeline
+     */
+    private $nextHandler;
+
+    /**
+     * @var RequestInterface Initial request with uncompressed body
+     */
+    private RequestInterface $originalRequest;
+
+    /**
+     * @param callable $nextHandler
+     * @param CompressionOption|null $compressionOption
+     */
+    public function __construct(callable $nextHandler, ?CompressionOption $compressionOption = null)
+    {
+        $this->nextHandler = $nextHandler;
+        $this->compressionOption = ($compressionOption) ?: new CompressionOption();
+    }
+
+    /**
+     * @param RequestInterface $request
+     * @param array $options
+     * @return PromiseInterface
+     */
+    public function __invoke(RequestInterface $request, array $options): PromiseInterface
+    {
+        $this->originalRequest = $request; // keep reference in case we have to retry with uncompressed body
+        if (!$this->shouldRetry($options)) {
+            $request = $this->compress($request);
+        }
+        $fn = $this->nextHandler;
+        return $fn($request, $options)->then(
+            $this->onFulfilled($options),
+            $this->onRejected($options)
+        );
+    }
+
+    /**
+     * if true, the request is retried with an uncompressed body
+     *
+     * @param array $options
+     * @return bool
+     */
+    private function shouldRetry(array $options): bool
```

Comment:
the name of this method is confusing compared with what it actually does. The condition should be reversed.

Reply:
Changed it. Please lmk if the new name communicates intent better.